jass/lib/DamageEvent.j

```diff
@@ -0,0 +1,675 @@
+library DamageEvent/* v 1.3.0.0
+**********************************************************************************
+*
+*   Physical Damage Detection Engine
+*   --------------------------------
+*   By looking_for_help aka eey
+*
+*   This system is able to detect, modify and deal damage. It can differentiate
+*   between physical and spell damage, as well as block, reduce or increase the
+*   applied damage to a desired value. You can also allocate damage events from
+*   running damage events.
+*
+**********************************************************************************
+*
+*   Implementation
+*   --------------
+*   1.  Copy this trigger to your map. With the AddDamageHandler function you can
+*       add as many handlers as you like (compare the OnDamage scope).
+*   2.  Copy the two custom abilities to your map and make sure they have the
+*       correct ID in the globals variable block.
+*   3.  Go to the locust swarm ability and invert its damage return portion
+*       from (default) 0.75 to -0.75.
+*   4.  Remove the spell damage reduction ability from the spell damage reduction
+*       items you use (runed bracers). You can configure the resistance of this
+*       item in the globals block, modifying BRACERS_SPELL_DAMAGE_REDUCTION.
+*
+**********************************************************************************
+*
+*   Important Notes
+*   ---------------
+*   1.  Life Drain does not work with this system, so you should use a triggered
+*       version of this spell if you want to use it.
+*   2.  Same for Finger of Death, if you want to use this spell bug free with this
+*       system, you should use a triggered version of it.
+*   3.  If you use damage modifiers by setting the damage amount variable, you have
+*       to use GetUnitLife, GetUnitMaxLife and SetUnitLife instead of GetWidgetLife,
+*       GetUnitState for UNIT_STATE_MAX_LIFE and SetWidgetLife in your whole map to
+*       ensure there occure no bugs.
+*   4.  The boolean USE_SPELL_RESISTANCE_AUTO_DETECT is only neccessary set to true
+*       if you want to use a customized damage table with spell damage resistance
+*       above 100%. If this is not the case, it should be set to false, as the
+*       system is faster then and still works correct.
+*   5.  As already mentioned you can't use the spell reduction ability when using this
+*       system (runed bracers and elunes grace). If you want to use them, you can
+*       trigger them by using the damage modifiers. Runed bracers is already considered
+*       in this system, so you don't have to code it.
```

Comment:
We use elunes grace. In order to merge this we'll have to change how the spear damage reduction library works.

cc @Aluvian-Sprinter